### PR TITLE
API: Add `Styler.to_html`, for saving output to HTML file

### LINKED ISSUE #11700: CLN/API: implemented to_html in terms of .style
Implement `to_html` / notebook repr based on `.style`.

prob need to expand this to take a `use` argument (to select the style, needs to be 'classic' for a while, to replicate the current `.to_html` one).

### LINKED ISSUE #13379: ENH: write Styler rendered output to file
I was answering a question on conditional formatting of html on [SO](http://stackoverflow.com/questions/37638402/pandas-html-output-conditional-formatting-highlight-cell-if-value-in-range/37647100?noredirect=1#comment62774932_37647100).

So the OP was using `to_html` but wanted conditional formatting. Therefore, I pointed to `Styler` to do this more easily than using the `formatters` argument in `to_html`, but when you have a `Styler` object, there is no longer a `to_html` method to write the result to a html file.

Of course you can write the result of `render` manually to a file, but would this ability be a desired enhancement?
- Maybe we can add a `to_file` method to `Styler`? (`to_html` seems a bit specific, if we also want to use the Styler for other formats such as latex in the future)
- Alternatively, we could add a `style` option in `to_html` where you can pass a Styler object?

Related to #11610

@TomAugspurger

### PR DESCRIPTION
**Pull request by attack68**
- [x] closes #11700 and closes #13379

## `Styler.to_html` method

This PR adds the above method to `Styler`. 

This is a **minimal** introduction with limited arguments to provide just the core file io functionality and basic options, either to write a fully structured HTML file or just the `Styler.render` output. It is very easier to extend in follow ons.

#### Jinja 2 template inheritance

Due to the more complex logic, the jinja2 templates were split into smaller components:

 - html_style.tpl: this just deals with the `<style>` html element
 - html_table.tpl: this just deals with the core `<table>` element including (or excluding) style identifiers
 - html.tpl: this is the controlling template which imports the above based on what is required.

It is more performant to have a separate temple that excludes styling elements by about 33%, rather than include the logic in a single template. I think it is easier to maintain as well.

#### Tests

A new `test_to_html.py` file is added. Some tests are moved from `test_style.py` if they more relate to HTML generation. Some new tests are added.

#### Objective

Ultimately the medium-term aim is to deprecate `DataFrame.to_html`. `Styler` is currently faster at rendering HTML than `DataFrame.to_html` (#39972), and this is further improved in this PR.
There is a standalone purpose for wanting to create HTML to file from Styler.
This PR complements `Styler.to_latex` whose aim is also to deprecate `DataFrame.to_latex`.

### CODE REVIEW

File: pandas/io/formats/style.py [hunk 1]
```diff
@@ -379,6 +382,67 @@ def to_excel(
...
+        table_attributes: str, optional
+            Attributes to assign within the `<table>` HTML element in the format:
+
+            ``<table .. <table_attributes> >``
+
+            If not given defaults to Styler's preexisting value.
+        encoding : str, optional
+            Character encoding setting for file output, and HTML meta tags,
+            defaults to "utf-8" if None.
+        doctype_html : bool, default True
```
Comment (jreback) [inline pandas/io/formats/style.py:416]: whats the utility of these last 2 options?

File: pandas/tests/io/formats/style/test_style.py [hunk 2]
```diff
@@ -1258,21 +1257,6 @@ def test_column_and_row_styling(self):
         )
         assert "#T__ .row0 {\n  color: blue;\n}" in s.render()
 
-    def test_colspan_w3(self):
```
Comment (jreback) [inline pandas/tests/io/formats/style/test_style.py:1261]: these are just moved?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/formats/style.py (hunk 1 above)
Comment (attack68) [reply pandas/io/formats/style.py:416]: `doctype_html` specifies whether to output a properly formed HTML document with appropriate `<html>`, `<body>` and `<head>` tags, or **just** the `<style>` and `<table>` tags.

`exclude_styles` provides an option in keeping with `DataFrame.to_html()`and it excludes all HTML `classes` and HTML `ids` and the `<style>` tag. (It is also more performant)

File: pandas/tests/io/formats/style/test_style.py (hunk 2 above)
Comment (attack68) [reply pandas/tests/io/formats/style/test_style.py:1261]: yes just moved to `test_html.py` since HTML related.

File: pandas/io/formats/style.py [hunk 3]
```diff
@@ -384,6 +387,68 @@ def to_excel(
...
+        buf: FilePathOrBuffer[str] | None = None,
+        *,
+        table_uuid: str | None = None,
+        table_attributes: str | None = None,
+        encoding: str | None = None,
+        doctype_html: bool = True,
+        exclude_styles: bool = False,
+    ):
+        """
+        Write Styler to a file, buffer or string in HTML-CSS format.
```
Comment (jreback) [inline pandas/io/formats/style.py:401]: add a versionadded 1.3

File: pandas/io/formats/style.py [hunk 4]
```diff
@@ -384,6 +387,68 @@ def to_excel(
...
+        exclude_styles : bool, default False
+            Whether to include the ``<style>`` element and all associated element
+            ``class`` and ``id`` identifiers, or solely the ``<table>`` element without
+            styling identifiers.
+
+        Returns
+        -------
+        str or None
+            If `buf` is None, returns the result as a string. Otherwise returns `None`.
+        """
```
Comment (jreback) [inline pandas/io/formats/style.py:434]: add a See Also back to DataFrame.to_html (or are we deprecating this in this PR?)

**Review by jreback — CHANGES_REQUESTED** [review]
great very minor comment, ping on green.

File: pandas/io/formats/style.py (hunk 3 above)
Comment (attack68) [reply pandas/io/formats/style.py:420]: done

File: pandas/io/formats/style.py (hunk 4 above)
Comment (attack68) [reply pandas/io/formats/style.py:459]: can't deprecate yet. working towards but needs a gap analysis: there are some things mising.

**Comment by attack68** [comment]
@jreback ping: greenish, failure unrelated

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @attack68

**Pull request merged**